streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -285,7 +285,11 @@ public StateStore getStore(final String name) {
         final Map<TopicPartition, Long> changelogOffsets = new HashMap<>();
         for (final StateStoreMetadata storeMetadata : stores.values()) {
             if (storeMetadata.changelogPartition != null) {
-                changelogOffsets.put(storeMetadata.changelogPartition, storeMetadata.offset);
+                // for changelog whose offset is unknown, use 0L indicating earliest offset
+                // otherwise return the current offset + 1 as the next offset to fetch
+                changelogOffsets.put(
```

Comment:
note the `offset` inside state store is the last record's offset, but to return `changelogOffsets` used for comparing with other end offsets etc we need to return `the next offset to process` (and yes it is still vulnerable to txn marker cases but that's fine for lag info I think).